src/main/scala-3/net/codingwell/scalaguice/ScalaModule.scala

```diff
@@ -0,0 +1,189 @@
+ *   }
+ * }
+ * }}}
+ * use
+ * {{{
+ * class MyModule extends AbstractModule with ScalaModule {
+ *   def configure {
+ *     bind[Service].to[ServiceImpl].in[Singleton]
+ *     bind[CreditCardPaymentService]
+ *     bind[Bar[Foo]].to[FooBarImpl]
+ *     bind[PaymentService].to[CreditCardPaymentService]
+ *
+ *     bindInterceptor[AOPI](methodMatcher = annotatedWith[AOP])
+ *   }
+ * }
+ * }}}
+ *
+ * '''Note''' This syntax allows binding to and from generic types.
+ * It doesn't currently allow bindings between wildcard types because the
+ * manifests for wildcard types don't provide access to type bounds.
+ */
+
+trait InternalModule[B <: Binder] {
+  import ScalaModule._
+
+  class BindingBuilder[T](typeLit: TypeLiteral[T]) extends ScalaAnnotatedBindingBuilder[T] {
+    val myBinder = binderAccess
+    val self = myBinder.bind(typeLit)
+  }
+
+  protected[this] def binderAccess: B
+
+  protected[this] inline def bind[T] = new BindingBuilder(typeLiteral[T])
+
+  protected[this] def bindInterceptor[I <: MethodInterceptor : ClassTag](classMatcher: Matcher[_ >: Class[_]] = Matchers.any(), methodMatcher: Matcher[_ >: AnnotatedElement]): Unit = {
+    val myBinder = binderAccess
+    val interceptor = implicitly[ClassTag[I]].runtimeClass.getDeclaredConstructor(Array.empty[Class[_]]: _*).newInstance().asInstanceOf[MethodInterceptor]
+    myBinder.requestInjection(interceptor)
+    myBinder.bindInterceptor(classMatcher, methodMatcher, interceptor)
+  }
+
+  protected[this] def annotatedWith[A <: Annotation : ClassTag]: Matcher[AnnotatedElement] = {
+    Matchers.annotatedWith(cls[A])
+  }
+
+  protected[this] def bindScope[T <: Annotation : ClassTag](scope: Scope): Unit = binderAccess.bindScope(cls[T], scope)
+  protected[this] def requestStaticInjection[T: ClassTag](): Unit = binderAccess.requestStaticInjection(cls[T])
+  protected[this] def getProvider[T: ClassTag]: Provider[T] = binderAccess.getProvider(cls[T])
+  protected[this] inline def getMembersInjector[T]: MembersInjector[T] = binderAccess.getMembersInjector(typeLiteral[T])
+}
+
+trait ScalaModule extends InternalModule[Binder] { self: AbstractModule =>
+  // Addresses http://lampsvn.epfl.ch/trac/scala/ticket/3564
+  // using Java reflection (as is done in Play's AkkaGuiceSupport)
+
+  import ScalaModule._
+
+  protected[this] def binderAccess: Binder = {
+    val method: Method = classOf[AbstractModule].getDeclaredMethod("binder")
+    if (!method.canAccess(this)) {
+      method.setAccessible(true)
+    }
+    method.invoke(this).asInstanceOf[Binder]
+      .withSource(filterTrace((new Throwable).getStackTrace))
+  }
+}
+
+trait ScalaPrivateModule extends InternalModule[PrivateBinder] { self: PrivateModule =>
+
+  import ScalaModule._
+
+  protected[this] def binderAccess: PrivateBinder = {
+    val method: Method = classOf[PrivateModule].getDeclaredMethod("binder")
+    if (!method.canAccess(this)) {
```

Comment:
JDK9 Dependancy
Scala3's min dependency is 8 which I would like to maintain to avoid unexpected confusion

Reply:
I'm gonna just remove the conditional. We ALWAYS will need to set accessible, so might as well do it without checking.